File: autocontext/src/autocontext/research/prompt_wiring.py
"""Research prompt wiring — format briefs for LLM injection (AC-501).

ResearchPromptInjector formats a ResearchBrief into a prompt section,
handling truncation to a char budget, confidence-based ordering, and
citation formatting. Supports placeholder injection or append-to-base.
"""

from __future__ import annotations

import logging

from autocontext.research.consultation import ResearchBrief

logger = logging.getLogger(__name__)

RESEARCH_PLACEHOLDER = "{research}"
DEFAULT_MAX_CHARS = 4000


class ResearchPromptInjector:
    """Formats research briefs and injects them into prompt templates."""

    def __init__(self, max_chars: int = DEFAULT_MAX_CHARS) -> None:
        self._max_chars = max_chars

    def format_brief(self, brief: ResearchBrief) -> str:
        """Render a brief as a markdown section, truncated to budget.

        Findings are ordered by confidence (highest first).
        Returns empty string if brief has no findings.
        """
        if not brief.findings:
            return ""

        sorted_findings = sorted(brief.findings, key=lambda f: f.confidence, reverse=True)

        parts: list[str] = [f"## External Research: {brief.goal}\n"]
        budget = self._max_chars - len(parts[0])

        for f in sorted_findings:
            block_lines = [f"**{f.query_topic}** (confidence: {f.confidence:.0%})"]
            block_lines.append(f.summary)
            for c in f.citations:
                if c.url:
                    block_lines.append(f"- [{c.source}]({c.url})")
                else:
                    block_lines.append(f"- {c.source}")
            block_lines.append("")
            block = "\n".join(block_lines)

            if len(block) > budget:
                if len(parts) == 1:
                    # At least include one truncated finding
                    parts.append(block[:budget])
                break
            parts.append(block)
            budget -= len(block)

        return "\n".join(parts)

    def inject(self, base_prompt: str, brief: ResearchBrief) -> str:
        """Inject formatted brief into a prompt template.

        If base_prompt contains {research}, replaces it.
        Otherwise appends the section after the base.
        Returns base_prompt unchanged if brief is empty.
        """
        section = self.format_brief(brief)
        if not section:
            return base_prompt

        if RESEARCH_PLACEHOLDER in base_prompt:
            return base_prompt.replace(RESEARCH_PLACEHOLDER, section)

        return f"{base_prompt}\n\n{section}"

File: ts/src/research/prompt-wiring.ts
/**
 * Research prompt wiring — format briefs for LLM injection (AC-501 TS parity).
 */

import { ResearchBrief } from "./consultation.js";

const PLACEHOLDER = "{research}";
const DEFAULT_MAX_CHARS = 4000;

export class ResearchPromptInjector {
  private maxChars: number;

  constructor(opts?: { maxChars?: number }) {
    this.maxChars = opts?.maxChars ?? DEFAULT_MAX_CHARS;
  }

  formatBrief(brief: ResearchBrief): string {
    if (!brief.findings.length) return "";

    const sorted = [...brief.findings].sort((a, b) => b.confidence - a.confidence);
    const header = `## External Research: ${brief.goal}\n`;
    const parts = [header];
    let budget = this.maxChars - header.length;

    for (const f of sorted) {
      const lines = [`**${f.queryTopic}** (confidence: ${Math.round(f.confidence * 100)}%)`];
      lines.push(f.summary);
      for (const c of f.citations) {
        lines.push(c.url ? `- [${c.source}](${c.url})` : `- ${c.source}`);
      }
      lines.push("");
      const block = lines.join("\n");

      if (block.length > budget) {
        if (parts.length === 1) parts.push(block.slice(0, budget));
        break;
      }
      parts.push(block);
      budget -= block.length;
    }
    return parts.join("\n");
  }

  inject(basePrompt: string, brief: ResearchBrief): string {
    const section = this.formatBrief(brief);
    if (!section) return basePrompt;
    if (basePrompt.includes(PLACEHOLDER)) return basePrompt.replace(PLACEHOLDER, section);
    return `${basePrompt}\n\n${section}`;
  }
}
